### DISCUSSION TITLE
File count limits

### DISCUSSION THREAD

**Discussion by codeninja75** [minio/minio#13491] [Q&A] [answered]
I'm trying to eval min.io for a use case we've got.  We've got large numbers (1 Billion +) of objects that are mostly small.  Avg size of 4k.  A very small % will be in the 1Mb+ range, biggest we've observed has been 200Mb.  And this is for one instance of our app.  As we add additional instances I could easily see us hitting 10s or 100s of billions.  Is there a document or blog post that talks about ram/cpu limits per some amt of objects?  Is there a recommended limit of number of objects per bucket?

**Comment by eco-minio** [comment, marked as answer]
Make sure to put your objects into a proper hierarchy - https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html.  Think of it the same way as a filesystem, you would not put billions of objects in a top level directory, so it is the same concept here. 
 
 Production hardware recommendations can be found here - https://min.io/product/reference-hardware.  CPU and RAM consumption is generally very low for MinIO.  More cores will typically be more beneficial than fewer, faster cores.  Keep in mind as well that CPU and RAM is tied more to active requests rather than total number of objects.  

 There is a scanner process that runs in the background that does some accounting functions like determining used space, whether an object needs to be tiered/replicated etc.  Billions of objects may take some time to crawl depending on the hardware, obviously faster drives helps here and having SSD/NVME will eliminate some of the overhead that would have been incurred from seeks and help with small file performance.  

In all cases, I/O should be the bottleneck for MinIO deployments (ideally at the network), so you should ensure you have sufficient network and disk speed to satisfy your applications requirements.  Some benchmark information here as well - https://min.io/resources#benchmarks

**Reply by codeninja75** [reply]
Thank you for the info!
